saffron/src/lib.rs

```diff
@@ -1359,194 +1360,128 @@ impl Cron {
-                [Sun, Mon, Tue, Wed, Thu, Fri, Sat],
-                [Mon, Tue, Wed, Thu, Fri, Sat, Sun],
-                [Tue, Wed, Thu, Fri, Sat, Sun, Mon],
-            ],
-            [
-                [Tue, Wed, Thu, Fri, Sat, Sun, Mon],
-                [Wed, Thu, Fri, Sat, Sun, Mon, Tue],
-                [Thu, Fri, Sat, Sun, Mon, Tue, Wed],
-                [Fri, Sat, Sun, Mon, Tue, Wed, Thu],
-                [Sat, Sun, Mon, Tue, Wed, Thu, Fri],
-                [Sun, Mon, Tue, Wed, Thu, Fri, Sat],
-                [Mon, Tue, Wed, Thu, Fri, Sat, Sun],
-            ],
-            [
-                [Mon, Tue, Wed, Thu, Fri, Sat, Sun],
-                [Tue, Wed, Thu, Fri, Sat, Sun, Mon],
-                [Wed, Thu, Fri, Sat, Sun, Mon, Tue],
-                [Thu, Fri, Sat, Sun, Mon, Tue, Wed],
-                [Fri, Sat, Sun, Mon, Tue, Wed, Thu],
-                [Sat, Sun, Mon, Tue, Wed, Thu, Fri],
-                [Sun, Mon, Tue, Wed, Thu, Fri, Sat],
-            ],
-        ];
-
-        TABLE[first_day_of_current_weekday][expected_first_day][current_weekday]
-    }
-
-    /// Gets the next matching (current inclusive) month that matches the cron expression.
-    /// The returned matching month is a value 0-11.
-    fn find_next_month(&self, current_month: u32) -> Option<u32> {
+    /// Gets the start of the next matching (current inclusive) month that matches the cron
+    /// expression.
+    fn find_next_month(&self, date: Date<Utc>) -> Option<Date<Utc>> {
         let Months(map) = self.months;
+        let current_month = date.month0();
         let bottom_cleared = (map >> current_month) << current_month;
-        let trailing_zeroes = bottom_cleared.trailing_zeros();
-        if trailing_zeroes < Months::BITS as u32 {
-            Some(trailing_zeroes)
+        let trailing_zeros = bottom_cleared.trailing_zeros();
+        if trailing_zeros < Months::BITS as u32 {
+            Utc.ymd_opt(date.year(), trailing_zeros + 1, 1).single()
         } else {
             None
         }
     }
 
-    fn find_next_time(&self, time: NaiveTime) -> Option<NaiveTime> {
-        let hour = time.hour();
-        if self.hours.contains_hour(time) {
-            if let Some(next_minute) = self.find_next_minute(time.minute()) {
-                return Some(NaiveTime::from_hms(hour, next_minute, 0));
+    fn find_next_date(
+        &self,
+        mut date: Date<Utc>,
+        bound: Date<Utc>,
+    ) -> Result<Option<Date<Utc>>, OutOfBound> {
+        if self.months.contains_month(date) {
+            match self.find_next_day(date) {
+                Some(next_day) if next_day > bound => return Err(OutOfBound),
+                Some(next_day) => return Ok(Some(next_day)),
+                None => {}
             }
         }
 
-        if hour < 23 {
-            if let Some(next_hour) = self.find_next_hour(hour + 1) {
-                let minute = self.find_next_minute(0).expect(Self::MUST_CONTAIN_MINUTE);
+        loop {
```

Comment:
nit: Using loop here makes the code very expressive. But I always get paranoid when I read code with `loop` because I know I need to very carefully check the termination conditions to make sure it doesn't infinite loop. But, we know this loop is eventually going to terminate, because `date` can't ever advance beyond a certain distance from `bound`. Is it possible to restructure this so it's a for loop instead? Possibly over some clever iterator that yields possible candidates for the next date. That way you can guarantee the code terminates.

Reply:
I don't think there's a clever way to make a for... Infinite loops should be protected against though through the `any()` check we do before making the iterator at all, so this shouldn't be called with a cron that can never match.